File: python/packages/core/src/matimo/encodings/parameter_encoding.py
"""
Parameter encoding — converts parameter groups into encoded string values.
Mirrors: packages/core/src/encodings/parameter-encoding.ts

Supported encoding types:
  mime_rfc2822_base64url  — Gmail API: RFC 2822 MIME message → base64url
  json_compact            — JSON.dumps compact equivalent
  url_encoded             — application/x-www-form-urlencoded
"""
from __future__ import annotations

import base64
import json
from typing import Any
from urllib.parse import urlencode

from matimo.errors import ErrorCode, MatimoError


def apply_parameter_encodings(
    params: dict[str, Any],
    encodings: list[dict[str, Any]],
) -> dict[str, Any]:
    """
    Apply a list of encoding configs to params, returning a new dict.
    Source parameters are consumed (removed) and replace with the target key.

    Args:
        params:    Original parameter dict.
        encodings: List of ParameterEncodingConfig-compatible dicts.

    Returns:
        New dict with encoded values written to target keys.
    """
    result = dict(params)

    for index, enc in enumerate(encodings):
        # Validate source
        source_keys_raw = enc.get("source")
        if not isinstance(source_keys_raw, list) or not source_keys_raw:
            raise MatimoError(
                f"Invalid parameter encoding config at index {index}: 'source' must be a non-empty list.",
                ErrorCode.INVALID_PARAMETER,
                {"index": index, "source": source_keys_raw},
            )
        if any(not isinstance(k, str) or not k for k in source_keys_raw):
            raise MatimoError(
                f"Invalid parameter encoding config at index {index}: all 'source' entries must be non-empty strings.",
                ErrorCode.INVALID_PARAMETER,
                {"index": index, "source": source_keys_raw},
            )
        source_keys: list[str] = source_keys_raw

        # Validate target
        target_key_raw = enc.get("target")
        if not isinstance(target_key_raw, str) or not target_key_raw:
            raise MatimoError(
                f"Invalid parameter encoding config at index {index}: 'target' must be a non-empty string.",
                ErrorCode.INVALID_PARAMETER,
                {"index": index, "target": target_key_raw},
            )
        target_key: str = target_key_raw

        # Validate encoding
        encoding_raw = enc.get("encoding")
        if not isinstance(encoding_raw, str) or not encoding_raw:
            raise MatimoError(
                f"Invalid parameter encoding config at index {index}: 'encoding' must be a non-empty string.",
                ErrorCode.INVALID_PARAMETER,
                {"index": index, "encoding": encoding_raw},
            )
        encoding: str = encoding_raw

        # Validate options
        options_raw = enc.get("options")
        if options_raw is not None and not isinstance(options_raw, dict):
            raise MatimoError(
                f"Invalid parameter encoding config at index {index}: 'options' must be a dict when provided.",
                ErrorCode.INVALID_PARAMETER,
                {"index": index, "options": options_raw},
            )
        options: dict[str, Any] = options_raw or {}

        # Gather source values
        source_values: dict[str, Any] = {k: result[k] for k in source_keys if k in result}

        encoded = _encode(source_values, encoding, options)

        # Remove source keys, write target
        for k in source_keys:
            result.pop(k, None)
        result[target_key] = encoded

    return result


# ---------------------------------------------------------------------------
# Internal dispatchers
# ---------------------------------------------------------------------------


def _encode(values: dict[str, Any], encoding: str, options: dict[str, Any]) -> str:
    if encoding == "mime_rfc2822_base64url":
        return _encode_mime_rfc2822(values)
    if encoding == "json_compact":
        return _encode_json_compact(values)
    if encoding == "url_encoded":
        return _encode_url_encoded(values)
    raise MatimoError(
        f"Unknown parameter encoding type: '{encoding}'",
        ErrorCode.INVALID_PARAMETER,
        {"encoding": encoding, "supported": ["mime_rfc2822_base64url", "json_compact", "url_encoded"]},
    )


def _encode_mime_rfc2822(values: dict[str, Any]) -> str:
    """
    Build an RFC 2822 MIME email and base64url-encode it.
    Used by the Gmail API 'send' tool.

    Expected keys: to, subject, body (plain text).
    """
    to = values.get("to", "")
    subject = values.get("subject", "")
    body = values.get("body", "")
    from_addr = values.get("from", "")
    cc = values.get("cc", "")
    bcc = values.get("bcc", "")

    lines: list[str] = []
    if from_addr:
        lines.append(f"From: {from_addr}")
    if to:
        lines.append(f"To: {to}")
    if cc:
        lines.append(f"Cc: {cc}")
    if bcc:
        lines.append(f"Bcc: {bcc}")
    lines.append(f"Subject: {subject}")
    lines.append("MIME-Version: 1.0")
    lines.append("Content-Type: text/plain; charset=utf-8")
    lines.append("Content-Transfer-Encoding: 7bit")
    lines.append("")
    lines.append(str(body))

    mime_bytes = "\r\n".join(lines).encode("utf-8")
    return base64.urlsafe_b64encode(mime_bytes).decode("ascii").rstrip("=")


def _encode_json_compact(values: dict[str, Any]) -> str:
    """JSON.stringify equivalent — compact, no whitespace."""
    if len(values) == 1:
        # Single source value — encode the value directly (not the wrapper dict)
        return json.dumps(next(iter(values.values())), separators=(",", ":"))
    return json.dumps(values, separators=(",", ":"))


def _encode_url_encoded(values: dict[str, Any]) -> str:
    """application/x-www-form-urlencoded encoding."""
    return urlencode({k: str(v) for k, v in values.items()})

File: typescript/packages/core/src/encodings/parameter-encoding.ts
/**
 * Generic parameter encoding system for transforming input parameters
 * Used to convert structured parameters into API-specific formats (MIME, JSON, etc.)
 *
 * This is YAML-driven via the parameter_encoding directive in tool definitions.
 * No tool-specific code needed - keeps Matimo universal.
 */

import { MatimoError, ErrorCode } from '../errors/matimo-error.js';

/**
 * Encoding configuration from YAML
 */
export interface ParameterEncodingConfig {
  source: string[]; // Parameter names to combine (e.g., [to, subject, body])
  target: string; // Parameter name to store result in (e.g., raw)
  encoding: string; // Encoding type (mime_rfc2822_base64url, json_compact, etc.)
  options?: Record<string, unknown>; // Optional config per encoding type
}

/**
 * Apply parameter encodings to transform input parameters
 * @param params - User-provided parameters
 * @param encodings - List of encoding configurations from tool YAML
 * @returns Parameters with encoded values added
 */
export function applyParameterEncodings(
  params: Record<string, unknown>,
  encodings: ParameterEncodingConfig[]
): Record<string, unknown> {
  const result = { ...params };

  for (const config of encodings) {
    // Extract source parameters
    const sourceValues: Record<string, unknown> = {};
    for (const key of config.source) {
      if (key in params) {
        sourceValues[key] = params[key];
      }
    }

    // Apply encoding function
    const encodedValue = encodeParameters(sourceValues, config.encoding, config.options);

    // Store in target parameter
    result[config.target] = encodedValue;
  }

  return result;
}

/**
 * Route to specific encoding function based on type
 */
function encodeParameters(
  sourceValues: Record<string, unknown>,
  encoding: string,
  _options?: Record<string, unknown>
): string {
  switch (encoding) {
    case 'mime_rfc2822_base64url':
      return encodeMimeRfc2822(sourceValues);
    case 'json_compact':
      return JSON.stringify(sourceValues);
    case 'url_encoded':
      return encodeUrlParams(sourceValues);
    default:
      throw new MatimoError(
        `Unknown parameter encoding type: ${encoding}`,
        ErrorCode.INVALID_PARAMETER,
        {
          encodingType: encoding,
          supportedTypes: ['mime_rfc2822_base64url', 'json_compact', 'url_encoded'],
        }
      );
  }
}

/**
 * Encode parameters as RFC 2822 MIME message with base64url encoding
 * Used for Gmail and other email APIs that accept raw MIME format
 */
function encodeMimeRfc2822(sourceValues: Record<string, unknown>): string {
  const to = sourceValues.to as string;
  const subject = sourceValues.subject as string;
  const body = sourceValues.body as string;
  const cc = sourceValues.cc as string | undefined;
  const bcc = sourceValues.bcc as string | undefined;
  // Accept both camelCase (`isHtml`) and snake_case (`is_html`) for backward compatibility
  const isHtml = ((sourceValues.isHtml ?? sourceValues.is_html) as boolean) || false;

  if (!to || !subject || !body) {
    throw new MatimoError(
      'MIME encoding requires: to, subject, body parameters',
      ErrorCode.INVALID_PARAMETER,
      {
        requiredFields: ['to', 'subject', 'body'],
        providedFields: Object.keys(sourceValues),
      }
    );
  }

  // Build RFC 2822 MIME message
  let message = '';
  message += `To: ${to}\r\n`;
  message += `Subject: ${subject}\r\n`;
  if (cc) message += `Cc: ${cc}\r\n`;
  if (bcc) message += `Bcc: ${bcc}\r\n`;
  message += `Content-Type: ${isHtml ? 'text/html' : 'text/plain'}; charset="UTF-8"\r\n`;
  message += 'Content-Transfer-Encoding: 7bit\r\n';
  message += '\r\n'; // Empty line separates headers from body
  message += body;

  // Convert to base64url
  const base64 = Buffer.from(message).toString('base64');
  // Replace standard base64 characters with URL-safe variants, then strip
  // trailing padding. Use while+endsWith instead of a regex to avoid ReDoS.
  let result = base64.replace(/\+/g, '-').replace(/\//g, '_');
  while (result.endsWith('=')) result = result.slice(0, -1);
  return result;
}

/**
 * Encode parameters as URL-encoded form data
 */
function encodeUrlParams(sourceValues: Record<string, unknown>): string {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(sourceValues)) {
    if (value !== undefined && value !== null) {
      params.append(key, String(value));
    }
  }
  return params.toString();
}

/**
 * Example usage in YAML:
 *
 * execution:
 *   type: http
 *   method: POST
 *   url: 'https://www.googleapis.com/gmail/v1/users/me/messages/send'
 *   headers:
 *     Content-Type: 'application/json'
 *     Authorization: 'Bearer {GMAIL_ACCESS_TOKEN}'
 *
 *   parameter_encoding:
 *     - source: [to, subject, body, cc, bcc, isHtml]
 *       target: raw
 *       encoding: mime_rfc2822_base64url
 *
 *   body:
 *     raw: '{raw}'
 *
 * Then users call:
 *   matimo.execute('gmail-send-email', {
 *     to: 'user@example.com',
 *     subject: 'Hello',
 *     body: 'Hi there',
 *     GMAIL_ACCESS_TOKEN: token
 *   });
 */
